src/Stack/ModuleInterface.hs

```diff
@@ -0,0 +1,425 @@
+module Stack.ModuleInterface
+    ( Interface(..)
+    , FastString(..)
+    , List(..)
+    , Dictionary(..)
+    , Module(..)
+    , Usage(..)
+    , Dependencies(..)
+    , getInterface
+    ) where
+
+{- HLINT ignore "Reduce duplication" -}
+
+import           Control.Monad   (replicateM, replicateM_, when)
+import           Data.Binary
+import           Data.Binary.Get (bytesRead, getInt64be, getWord32be,
+                                  getWord64be, getWord8, lookAhead, skip)
+import           Data.Bool       (bool)
+import           Data.Char       (chr)
+import           Data.Functor    (void, ($>))
+import           Data.List       (find)
+import           Data.Maybe      (catMaybes)
+import           Data.Semigroup  ((<>))
+import qualified Data.Vector     as V
+import           Foreign         (sizeOf)
+import           Numeric         (showHex)
+
+type IsBoot = Bool
+
+type ModuleName = FastString
+
+newtype List a = List
+    { unList :: [a]
+    } deriving (Show)
+
+newtype FastString = FastString
+    { unFastString :: String
+    } deriving (Show)
+
+newtype Dictionary = Dictionary
+    { unDictionary :: V.Vector FastString
+    } deriving (Show)
+
+newtype Module = Module
+    { unModule :: ModuleName
+    } deriving (Show)
+
+newtype Usage = Usage
+    { unUsage :: FilePath
+    } deriving (Show)
+
+data Dependencies = Dependencies
+    { dmods    :: List (ModuleName, IsBoot)
+    , dpkgs    :: List (ModuleName, Bool)
+    , dorphs   :: List Module
+    , dfinsts  :: List Module
+    , dplugins :: List ModuleName
+    } deriving (Show)
+
+data Interface = Interface
+    { deps  :: Dependencies
+    , usage :: List Usage
+    } deriving (Show)
+
+-- | Read a block prefixed with its length
+withBlockPrefix :: Get a -> Get a
+withBlockPrefix f = getWord32be *> f
+
+getBool :: Get Bool
+getBool = toEnum . fromIntegral <$> getWord8
+
+getString :: Get String
+getString = fmap (chr . fromIntegral) . unList <$> getList getWord32be
+
+getMaybe :: Get a -> Get (Maybe a)
+getMaybe f = bool (pure Nothing) (Just <$> f) =<< getBool
+
+getList :: Get a -> Get (List a)
+getList f = do
+    i <- getWord8
+    l <-
+        if i == 0xff
+            then getWord32be
+            else pure (fromIntegral i :: Word32)
+    List <$> replicateM (fromIntegral l) f
+
+getTuple :: Get a -> Get b -> Get (a, b)
+getTuple f g = (,) <$> f <*> g
+
+getFastString :: Get FastString
```

Comment:
Any reason to not use a `ByteString` here?

Reply:
No particular reason at all, I will change it.